### ISSUE TITLE
Incorrect conditional requests strategy: etag/last-modified must be updated/persisted even if content-checksum is unchanged

### ISSUE THREAD

**Issue by bensheldon** [feedbin/feedbin#726]
A blog I follow (via Feedbin ❤ ) has been collecting data about [RSS feedreaders' bad behavior](https://rachelbythebay.com/w/2024/08/02/fs/). Feedbin is mentioned in the list:

> Feedbin. Hella broken IMS/INM caching, terrible pacing, regular forced unconditional polls, etc. They trip 429s in prod constantly. There are a bunch of these and they're all broken the same way, so I chalk that up to the program/service and not any one user's config.

I took a look to see if this had been mentioned in Feedbin's Issues previously and didn't see it, so I hope this isn't a duplicate and is new information.

In the most recent blog post (there's a whole series) ["A common bug in a bunch of feed readers" ](https://rachelbythebay.com/w/2024/08/17/hash/) it describes:

> [The request library will] do some kind of "hash match" thing based on the body of the feed - like a md5sum or somesuch.
> 
> It goes something like this, apparently: fetch the feed for the first time. Then store the hash, the Last-Modified time, and the ETag.
> 
> Next time, send If-Modified-Since and If-None-Match using the stored values. But, then, if the hash of the body matches, return immediately and don't update anything else... *even if* the Last-Modified and/or ETag values changed! So, next time, they send the old values again, and it never gets any better.

This has been fixed in several other feedreaders; example: https://inessential.com/2024/08/03/netnewswire_and_conditional_get_issues.html

**I think Feedbin does this similarly incorrectly too.**

Inside of Feedbin, it will make a web request with etag/last-modified values in the If-Modified-Since / If-None-Match headers and only parse if not modified:

https://github.com/feedbin/feedbin/blob/9dd0706656840619d810d514db89f980660972a2/app/jobs/feed_crawler/downloader.rb#L35-L42

`Feedkit::Request#not_modified?` [implementation exposes the problem described in the blog](https://github.com/feedbin/feedkit/blob/944b58acd5322a1c384cbfcce8c607341308bca3/lib/feedkit/response.rb#L36)`:

```ruby
def not_modified?(old_checksum = nil)
  status == 304 || old_checksum == checksum
end
```

My reading of Feedbin/Feedkit's implementation is that it does the same incorrect strategy described in the blog post: Feedbin only updates/persists tags/last-modified if the content changes, when instead, it should update them _when the server changes them_. The impact is that Feedbin will subsequently send effectively unconditional requests.

## What should Feedbin do instead?

I think that `Feedkit::Request` should differentiate between these two situations:

- "The server sent a 304 in response to our conditional request". in which case nothing needs to be updated.
- "The body/content of the response has changed", in which case the `last_modified` and `etag` should still be updated/persisted, even if the body/content parsing can be skipped.

I'll offer a partial solution here if it helps, I think `Feedkit::Request` could expose:

- `Feedkit::Request#conditionally_modified?` to distinguish whether the server returned a 304
- `Feedkit::Request#content_modified?` to distinguish whether the content itself has changed (and thus should parsed)

And then I think that the Downloader could be modified so that crawl_data is persisted in either scenario, regardless of whether the feed content is parsed or not:

https://github.com/feedbin/feedbin/blob/9dd0706656840619d810d514db89f980660972a2/app/jobs/feed_crawler/downloader.rb#L15-L19

**Comment by bensheldon** [comment]
Also, I want to flag that I think there's something bugged with `ignore_http_caching` too, because there is both `Downloader#ignore_http_caching?` and `CrawlData#ignore_http_caching?` and they are not synchronized with each other e.g. if a `critical` parse ignores HTTP caching, it won't update that value in the CrawlData (also, the logging in Downloader is slightly off as to whether it is truly unconditional or not).  And I think that it could exacerbate this problem of lots of unconditional fetches.

- https://github.com/feedbin/feedbin/commit/90fab3c13fad14a07dff202826b43c5c9b8ce94e
- https://github.com/feedbin/feedbin/commit/83306bc61241d2aa7fdf3052dfe4cc41be4862c5
- https://github.com/feedbin/feedbin/commit/d92031dce87ad8b6150952ec754243bbd775e2d2 <- this reverted part of the previous changes and made them more out of sync; I'm not sure about the context of that.

**Comment by benubois** [comment]
Resolved by #727.
